### ISSUE TITLE
Proposal: Helper to make sorting link generation easier

### ISSUE THREAD

**Issue by NickCraver** [aspnet/Mvc#7367]
A common scenario I run into is needing to use the current query string (action parameters) and change 1 for the purposes of sorting or filtering (e.g. a search query against the current other filters). Let's take an example:
```
/path/page?minRuns=20&filter=someText&database=PRIZM&sort=RunsPerSecond&order=Asc
```
In the view, I have a table with columns with headers that are sorting links.  Let's say I have 3 columns so the headers look like this:
```html
<th><a href="/path/page?minRuns=20&amp;filter=someText&amp;database=PRIZM&amp;sort=Runs">Runs</a></th>
<th><a href="/path/page?minRuns=20&amp;filter=someText&amp;database=PRIZM&amp;sort=Duration">Duration</a></th>
<!-- repeat 8 times... -->
```
Currently, generating the query string of a thing that *differs by one* is something users have to do very manually, generally with a builder or making a collection to serialize per link in a helper. I think this could be made much cleaner as something the framework provides. You can *almost* do this today with [`asp-all-route-data` via tag helper](https://docs.microsoft.com/en-us/aspnet/core/mvc/views/tag-helpers/built-in/anchor-tag-helper#asp-all-route-data), but it could be improved.

Here's the most succinct way I can find to do this today (note this is just query params, but you get the idea):
```html
@{ var routeDict = Context.Request.Query.ToDictionary(q => q.Key, q => q.Value.ToString()); }
<th><a asp-all-route-data="routeDict" asp-route-sort="@Sorts.Runs">Runs</a></th>
<th><a asp-all-route-data="routeDict" asp-route-sort="@Sorts.Duration">Duration</a></th>
```

## Proposal
What if the anchor tag helper allowed `asp-all-route-data` without a value? The default would be the current route data. Since this already works in combination with `asp-route-{key}="value"`, the combination becomes simply:
```html
<th><a asp-all-route-data asp-route-sort="@Sorts.Runs">Runs</a></th>
<th><a asp-all-route-data asp-route-sort="@Sorts.Duration">Duration</a></th>
```

Perhaps there's a simpler and/or different way to achieve this (e.g. a new attribute), I was just thinking minimal changes would save code and open up use cases here. Thoughts?

/cc @DamianEdwards

**Comment by mkArtakMSFT** [comment]
@DamianEdwards any thoughts?

**Comment by DamianEdwards** [comment]
I agree this is a fairly common scenario and the solutions today are cumbersome. I also like the proposal of overloading the existing `asp-all-route-data` when used with no value to mean "give me all the current route params". This seems logical and I believe would be non-breaking.

Interested in @rynowak and @pranavkm 's thoughts. If this is fairly cheap, I'd consider it a candidate for 2.1.0-preview2 (to be weighed against everything else of course).

**Comment by pranavkm** [comment]
@NTaylorMullen should be able to correct me, but afaik minimized attributes only work on boolean property types. In this case, the property backing `asp-route-data` would be a dictionary. Supporting this might require changes in the parser, making it perhaps not too cheap.

**Comment by DamianEdwards** [comment]
Yet another case to support attribute overloading on Tag Helpers, sigh. One day.

**Comment by rynowak** [comment]
A simple or cheap way to do this would be to add a new parameter like `asp-add-route-data`. It's not exactly what's being proposed here, but it doesn't require a language change, and I personally like it better, because it's semantically distinct from the things you can do today (`add` is in the name).

I'm down on language changes for this kind thing unless we're interested in take a bigger look at how key-value-pairs work across all tag helpers. As proposed, this idea enshrines the concept of 'ambient values' in the language. I think we need to be careful about that.

**Comment by bgribaudo** [comment]
Related: #3791

**Comment by mkArtakMSFT** [comment]
We should review the ask in the referenced issue #3791 when we'll be deciding on the approach to take.

**Comment by tuespetre** [comment]
What about a boolean `asp-current-route-data` attribute that serves the same purpose as a 'boolean overload' of `asp-all-route-data`? The behavior of the tag helper would then be to apply the current values first if `asp-current-route-data` is `true`, then the `asp-all-route-data` values, then the keyed `asp-route-*` values. This way there is a nice 'layering' of functionality and everyone wins. 🥇 🥈 🥉 🏅
